orchagent/nhgorch.cpp

```diff
@@ -0,0 +1,1192 @@
+ *              ref counters.
+ * Params:      None.
+ * Returns:     Nothing.
+ */
+void NextHopGroupMember::remove()
+{
+    SWSS_LOG_ENTER();
+
+    /*
+     * If the member is already removed, exit so we don't decrement the ref
+     * counters more than once.
+     */
+    if (!isSynced())
+    {
+        return;
+    }
+
+    m_gm_id = SAI_NULL_OBJECT_ID;
+    gCrmOrch->decCrmResUsedCounter(CrmResourceType::CRM_NEXTHOP_GROUP_MEMBER);
+    gNeighOrch->decreaseNextHopRefCount(m_nh_key);
+}
+
+/*
+ * Purpose:     Destructor.
+ * Description: Assert the group member is removed and remove the labeled
+ *              next hop from NeighOrch if it is unreferenced.
+ * Params:      None.
+ * Returns:     Nothing.
+ */
+NextHopGroupMember::~NextHopGroupMember()
+{
+    SWSS_LOG_ENTER();
+
+    /*
+     * The group member should be removed from its group before destroying it.
+     */
+    assert(!isSynced());
+
+    /*
+     * If the labeled next hop is unreferenced, remove it from NeighOrch as
+     * NhgOrch and RouteOrch are the ones controlling it's lifetime.  They both
+     * watch over these labeled next hops, so it doesn't matter who created
+     * them as they're both doing the same checks before removing a labeled
+     * next hop.
+     */
+    if (isLabeled() &&
+        gNeighOrch->hasNextHop(m_nh_key) &&
+        (gNeighOrch->getNextHopRefCount(m_nh_key) == 0))
+    {
+        gNeighOrch->removeMplsNextHop(m_nh_key);
+    }
+}
+
+/*
+ * Purpose:     Constructor.
+ * Description: Initialize the group's members based on the next hop group key.
+ * Params:      IN  key - The next hop group's key.
+ * Returns:     Nothing.
+ */
+NextHopGroup::NextHopGroup(const NextHopGroupKey& key) :
+    m_key(key),
+    m_id(SAI_NULL_OBJECT_ID),
+    m_is_temp(false)
+{
+    SWSS_LOG_ENTER();
+
+    /* Parse the key and create the members. */
+    for (const auto& it : m_key.getNextHops())
+    {
+        m_members.emplace(it, NextHopGroupMember(it));
+    }
+}
+
+/*
+ * Purpose:     Move constructor.
+ * Description: Initialize the members by doing member-wise move construct.
+ * Params:      IN  nhg - The rvalue object to initialize from.
+ * Returns:     Nothing.
+ */
+NextHopGroup::NextHopGroup(NextHopGroup&& nhg) :
+    m_key(std::move(nhg.m_key)),
+    m_id(std::move(nhg.m_id)),
+    m_members(std::move(nhg.m_members)),
+    m_is_temp(nhg.m_is_temp)
+{
+    SWSS_LOG_ENTER();
+
+    /* Invalidate the rvalue SAI ID. */
+    nhg.m_id = SAI_NULL_OBJECT_ID;
+}
+
+/*
+ * Purpose:     Move assignment operator.
+ * Description: Perform member-wise swap.
+ * Params:      IN  nhg - The rvalue object to swap with.
+ * Returns:     Referene to this object.
+ */
+NextHopGroup& NextHopGroup::operator=(NextHopGroup&& nhg)
+{
+    SWSS_LOG_ENTER();
+
+    std::swap(m_key, nhg.m_key);
+    std::swap(m_id, nhg.m_id);
+    std::swap(m_members, nhg.m_members);
+    m_is_temp = nhg.m_is_temp;
+
```

Comment:
Does this remove the "nexthopgroup" object from SDK?

Reply:
As long as the object that is referenced by the rvalue argument is going to be destroyed (either because it's pointing to a temporary object like in the case at line `1031`, or because the object reaches its end of lifetime), yes, because the the destructor will be invoked on the rvalue-referenced object which will now contain the properties of the previous NHG on which the move assignment operator was invoked on.